core/state/state_object.go

```diff
@@ -196,9 +200,12 @@ func (s *stateObject) GetCommittedState(key common.Hash) common.Hash {
 	)
 
 	if s.db.witness != nil {
-		// when collecting witness data with snapshot enabled, always prefetch every read key to
-		// ensure that read storage slots will end up in the witness
-		s.db.prefetcher.prefetch(s.addrHash, s.data.Root, s.address, [][]byte{key[:]})
+		if !s.createdBlock && s.origin.Root != types.EmptyRootHash {
```

Comment:
@rjl493456442 @holiman previous to this commit, selfdestructing+recreating an account would cause the prefetcher to load the storage keys using the current root (always empty post byzantium).  Nothing would be loaded.

The error wasn't testable, and this commit is mostly cosmetic to make the logic around invocation of the prefetcher for storage reads more correct.

Reply:
I don't see how it can be correct to add an unjournalled flag to stateobjects though. I mean, if we add some more 'stateful thing' on stateobjects, then surely that stateful-ness must be journalled?

Reply:
How do you mean it's "mostly cosmetic" ? If it's "mostly cosmetic", it seems like an extremely high price to pay for a cosmetic change (because another created-flag on the stateobject makes a complicated thing even more complicated)

Reply:
> I don't see how it can be correct to add an unjournalled flag to stateobjects though. I mean, if we add some more 'stateful thing' on stateobjects, then surely that stateful-ness must be journalled?

oof. yeah.

> How do you mean it's "mostly cosmetic" ?

The previous code worked without error because a state object's root is only modified (set to `0x00..0`)during block processing if it is deleted and resurrected.  So it just means that the prefetcher would stop retrieving newly read storage keys for the account, which is the correct behavior.

The changes in this commit aim to make the code more intuitive.  But I guess adding overhead and complexity to statedb isn't worth it.  I'd opt to keep the old logic and add a comment blurb about why it is correct.